src/platforms/common/px4_work_queue/WorkQueueManager.cpp

```diff
@@ -0,0 +1,320 @@
+
+#include <lib/drivers/device/Device.hpp>
+
+using namespace time_literals;
+
+namespace px4
+{
+
+List<WorkQueue *> px4_work_queues_list;
+pthread_mutex_t px4_work_queues_list_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+BlockingQueue<wq_config *, 1> px4_wq_manager_new;
+
+static WorkQueue *
+find_work_queue(const char *name)
+{
+	pthread_mutex_lock(&px4_work_queues_list_mutex);
+
+	// search list
+	for (WorkQueue *wq = px4_work_queues_list.getHead(); wq != nullptr; wq = wq->getSibling()) {
+		if (strcmp(wq->get_name(), name) == 0) {
+			pthread_mutex_unlock(&px4_work_queues_list_mutex);
+			return wq;
+		}
+	}
+
+	pthread_mutex_unlock(&px4_work_queues_list_mutex);
+
+	return nullptr;
+}
+
+static void add_workqueue(WorkQueue *wq)
+{
+	// add to work queue list
+	pthread_mutex_lock(&px4_work_queues_list_mutex);
+	px4_work_queues_list.add(wq);
+	pthread_mutex_unlock(&px4_work_queues_list_mutex);
+}
+
+static void remove_workqueue(WorkQueue *wq)
+{
+	// remove from work queue list
+	pthread_mutex_lock(&px4_work_queues_list_mutex);
+	px4_work_queues_list.remove(wq);
+	pthread_mutex_unlock(&px4_work_queues_list_mutex);
+}
+
+WorkQueue *
+work_queue_create(const wq_config &new_wq)
+{
+	// search list for existing work queue
+	WorkQueue *wq = find_work_queue(new_wq.name);
+
+	// create work queue if it doesn't exist
+	if (wq == nullptr) {
+
+		// add to list
+		// main thread wakes up, creates the thread
+		px4_wq_manager_new.push((wq_config *)&new_wq);
+
+		// we wait until new wq is created, then return
+		uint64_t t = 0;
+		wq = nullptr;
+
+		while (wq == nullptr && t < 10_s) {
+			wq = find_work_queue(new_wq.name);
+
+			// Wait up to 1s
+			t += 2500;
+			px4_usleep(2500);
+		}
+
+		if (wq == nullptr) {
+			PX4_ERR("work queue: %s failed to create", new_wq.name);
+
+		} else {
+			// work queue found, set PID
+			//wq->set_task_id(pid);
+		}
+	}
+
+	return wq;
+}
+
+const wq_config &device_bus_to_wq(uint32_t device_id_int)
+{
+	union device::Device::DeviceId device_id;
+	device_id.devid = device_id_int;
+
+	const device::Device::DeviceBusType bus_type = device_id.devid_s.bus_type;
+	const uint8_t bus = device_id.devid_s.bus;
+
+	if (bus_type == device::Device::DeviceBusType_I2C) {
+		if (bus == 1) {
+			return wq_configurations[I2C1];
+
+		} else if (bus == 2) {
+			return wq_configurations[I2C2];
+		}
+
+	} else if (bus_type == device::Device::DeviceBusType_SPI) {
+		if (bus == 1) {
+			return wq_configurations[SPI1];
+
+		} else if (bus == 2) {
+			return wq_configurations[SPI2];
+		}
+	}
+
+	// otherwise use high priority
+	return wq_configurations[hp_default];
+};
+
+
+
+/*****************************************************************************
+ *
+ *
+ *
+ *
+ */
+void
+WorkQueueManager::run()
+{
+	while (!should_exit()) {
+
+		// create new work queues as needed
+		wq_config *new_wq_config = px4_wq_manager_new.pop();
+
+		PX4_INFO("New WQ: %s priority: %d stack: %d", new_wq_config->name, SCHED_PRIORITY_MAX - new_wq_config->priority,
```

Comment:
Wrong sign.
![selection_015](https://user-images.githubusercontent.com/37091262/51642931-1308af80-1f28-11e9-96a2-030460681b62.png)

```suggestion
		PX4_INFO("New WQ: %s priority: %d stack: %d", new_wq_config->name, SCHED_PRIORITY_MAX + new_wq_config->priority,
```

Reply:
Ha, thanks.

Reply:
Making them relative was a recent lazy change to deal with (or not have to deal with) the other platforms.